Migrate everything from Python 2.7 to Python 3.6

bkgoksel opened the pull request:
As pip has recently told us in all caps:

`DEPRECATION: Python 2.7 will reach the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 won't be maintained after that date. A future version of pip will drop support for Python 2.7`

This is the beginning of a long, uncertain journey that we nevertheless must take.
The first step is to run `2to3` on our entire codebase and hope that the world doesn't burn around us.

Godspeed,

bkgoksel committed:
Use py 3.6 for now

bkgoksel committed:
Change non-code files to python 3.6

bkgoksel committed:
Remove futures from requirements

bkgoksel committed:
Switch to mysqlclient as Mysql-Python hasn't been updated in 5 years and isn't py3 compatible

bkgoksel committed:
Fix travis file syntax

bkgoksel committed:
Pass explicit bytes objects to base64encode

bkgoksel committed:
Don't base64 encode when you're going to concat with str anyways

bkgoksel committed:
Change the data in all urllib requests to be UTF-8 encoded

bkgoksel commented:
Known issue:
Python3 doesn't have the `err.message` attribute for exceptions, need to just pass the error itself as its string representation wherever this happens ie `err.message` -> `str(err)`

epicfaace committed:
Merge

epicfaace committed:
fix: replace e.message with str(e)

epicfaace committed:
fix: convert keys() iterator to list

epicfaace committed:
fix: switch to 3.6 because 3.7 doesn't work on ubuntu trusty

epicfaace committed:
fix: change other travis script to use 3.6

epicfaace committed:
fix: install python 3.6 on docker files too

epicfaace committed:
fix: add python3.6 shebang to codalab_service.py

epicfaace committed:
fix: use python 3 pip in setup.sh

epicfaace committed:
fix: use mysqlclient instead of MySQL-python

epicfaace committed:
fix: more python3 fixes in docker files

epicfaace committed:
readd python2 to default cpu/gpu, alias python 3

epicfaace committed:
fix: fix shebang

epicfaace committed:
fix: install python3-venv

epicfaace committed:
fix: install python-virtualenv

epicfaace committed:
fix: install python3-dev

epicfaace committed:
fix: install python3.6-dev

epicfaace committed:
fix: fix setup.sh

epicfaace committed:
change python version to 3

epicfaace committed:
fix: fix worker dockerfile

epicfaace committed:
fix: use python 3.6 for codalab_service

epicfaace committed:
fix: try python3 prefix

epicfaace committed:
use python3 instead of python

epicfaace committed:
fix: decode file after reading

epicfaace committed:
fix: run 2to3 again

epicfaace committed:
fix: print CalledProcessError

epicfaace committed:
fix: print error better

epicfaace committed:
fix: use check_output instead of check_call

epicfaace committed:
fix: fix worker dockerfile

epicfaace committed:
fix: fix base64 encode/decode issues

epicfaace committed:
fix: remove decode

epicfaace committed:
fix: fix encoding

epicfaace committed:
Update test_cli.py

epicfaace committed:
fix: remove print for now

epicfaace committed:
retrigger travis job

epicfaace committed:
fix: replace assertItemsEqual with assertEqual

epicfaace committed:
fix: add encode() for hashlib.sha1 and update

epicfaace committed:
fix: use  BytesIO instead of StringIO for gzip

epicfaace committed:
fix: fix more string / bytes issues

epicfaace committed:
fix: fix more unicode issues

epicfaace committed:
fix: sort os.listdir

epicfaace committed:
fix: use binary string

epicfaace committed:
fix: don't encode bytestring

epicfaace committed:
fix: decode gzip strings

epicfaace committed:
fix: return unicode from un_gzip_string

epicfaace committed:
fix: remove extraneous encode

epicfaace committed:
fix: use str instead of exception.message

epicfaace committed:
fix: decode httpresponse

epicfaace committed:
fix: use utf-8 when opening subprocess

epicfaace committed:
fix: fix contains_unicode

epicfaace committed:
fix: remove extraneous encode()

epicfaace committed:
fix: fix write string / write bytes

epicfaace committed:
fix: binary read for bz2 files

epicfaace committed:
fix: fix contains_unicode function

epicfaace committed:
fix: use integer division

epicfaace committed:
fix: use utf-8 encoding for subprocess.check_output in run_command

epicfaace committed:
fix: print args properly

epicfaace committed:
fix: fix sanitize method

epicfaace committed:
fix: fix contains_unicode

epicfaace committed:
fix: capture stderr in check_output

epicfaace committed:
fix: disable output truncating for now

epicfaace committed:
fix: print json_bundle for debugging

epicfaace committed:
fix: more helpful error message

epicfaace committed:
fix: fix unicode_constructor

epicfaace committed:
fix: fix unicode in text

epicfaace committed:
fix: another encode/decode issue

epicfaace committed:
fix: decode error_body

epicfaace committed:
fix: send byte strings

epicfaace committed:
fix: decode stringio error

epicfaace committed:
fix: print exceptions on command line

epicfaace committed:
fix: don't decode response

epicfaace committed:
fix: use wb for upload

epicfaace committed:
fix: try conditional encode/decode

epicfaace committed:
fix: rename python to python3

epicfaace committed:
Merge

epicfaace committed:
fix typo

epicfaace committed:
fix: don't encode data

epicfaace committed:
Revert "fix: don't encode data"

This reverts commit 2a6abb615ff8cdedd45db32b911a4cf2a26fd6e0.

epicfaace committed:
fix: go back to use w for upload

epicfaace committed:
fix: remove unnecessary check

epicfaace committed:
fix: encode blob before sql insert

epicfaace committed:
fix: conditionally decode error body

epicfaace committed:
fix: change encode_str and decode_str

epicfaace committed:
fix: don't capture stderr in check_output

epicfaace committed:
fix: replace getheader -> get

epicfaace committed:
fix: remove format_exc

epicfaace committed:
fix: accept both string and binary here

epicfaace committed:
fix: fix len(to_send)

epicfaace committed:
fix: make sure upload_to_bundle_store only writes to binary files

epicfaace committed:
fix: properly print binary contents to stdout.buffer

epicfaace committed:
fix: black format

epicfaace committed:
use xenial

epicfaace committed:
Revert "use xenial"

This reverts commit 7842bde12a54c84a9621a4cffccc61445cccdcbf.

epicfaace committed:
try flushing stdout

epicfaace committed:
fix: fix binary in worker_model

epicfaace committed:
fix: remove unnecessary decode

epicfaace committed:
feat: set max request params to 0 by default (python 2 allows comparing numbers with None but not python 3)

epicfaace committed:
fix: load json string

epicfaace committed:
ci: try changing travis script

epicfaace committed:
fix: show errors on test

epicfaace committed:
fix: fix typo

epicfaace committed:
Revert "ci: try changing travis script"

This reverts commit 2394d1d3ae27446437a8f753a6f0ad64014ed992.

epicfaace committed:
fix: add try-catch to checkin

epicfaace committed:
fix: json.dump to string-writable file

epicfaace committed:
fix: open files in rb mode for path_contents

epicfaace committed:
fix (tests): create binary versions of path_contents and run_command

epicfaace committed:
fix(test): more binary output fixes

epicfaace committed:
fix: properly encode POST data

epicfaace committed:
fix: add charset=utf8 to all db urls

epicfaace commented:
I kept getting this issue (when running `worksheets` tests):

```
>> codalab/bin/cl wedit 0xf51f5551139042c78fa3534dc6e19187 --file /home/epicfaace/codalab-worksheets/tests/files/unicode-worksheet
Traceback (most recent call last):
  File "/home/epicfaace/codalab-worksheets/codalab/client/json_api_client.py", line 16, in wrapper
    return f(*args, **kwargs)
  File "/home/epicfaace/codalab-worksheets/codalab/client/json_api_client.py", line 680, in update_worksheet_raw
    data='\n'.join(lines).encode(),
  File "/home/epicfaace/codalab-worksheets/worker/codalabworker/rest_client.py", line 91, in _make_request
    with closing(urllib.request.urlopen(request)) as response:
  File "/usr/lib/python3.6/urllib/request.py", line 223, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/lib/python3.6/urllib/request.py", line 532, in open
    response = meth(req, response)
  File "/usr/lib/python3.6/urllib/request.py", line 642, in http_response
    'http', request, response, code, msg, hdrs)
  File "/usr/lib/python3.6/urllib/request.py", line 570, in error
    return self._call_chain(*args)
  File "/usr/lib/python3.6/urllib/request.py", line 504, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.6/urllib/request.py", line 650, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 500: Internal Server Error
During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "codalab/bin/../../codalab/bin/cl.py", line 125, in <module>
    main()
  File "codalab/bin/../../codalab/bin/cl.py", line 119, in main
    cli.do_command(sys.argv[1:])
  File "/home/epicfaace/codalab-worksheets/codalab/lib/bundle_cli.py", line 848, in do_command
    structured_result = command_fn()
  File "/home/epicfaace/codalab-worksheets/codalab/lib/bundle_cli.py", line 842, in <lambda>
    command_fn = lambda: args.function(self, args)
  File "/home/epicfaace/codalab-worksheets/codalab/lib/bundle_cli.py", line 2932, in do_wedit_command
    client.update_worksheet_raw(worksheet_info['id'], lines)
  File "/home/epicfaace/codalab-worksheets/codalab/client/json_api_client.py", line 38, in wrapper
    sys.exc_info()[2],
  File "/home/epicfaace/codalab-worksheets/venv/lib/python3.6/site-packages/six.py", line 692, in reraise
    raise value.with_traceback(tb)
  File "/home/epicfaace/codalab-worksheets/codalab/client/json_api_client.py", line 16, in wrapper
    return f(*args, **kwargs)
  File "/home/epicfaace/codalab-worksheets/codalab/client/json_api_client.py", line 680, in update_worksheet_raw
    data='\n'.join(lines).encode(),
  File "/home/epicfaace/codalab-worksheets/worker/codalabworker/rest_client.py", line 91, in _make_request
    with closing(urllib.request.urlopen(request)) as response:
  File "/usr/lib/python3.6/urllib/request.py", line 223, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/lib/python3.6/urllib/request.py", line 532, in open
    response = meth(req, response)
  File "/usr/lib/python3.6/urllib/request.py", line 642, in http_response
    'http', request, response, code, msg, hdrs)
  File "/usr/lib/python3.6/urllib/request.py", line 570, in error
    return self._call_chain(*args)
  File "/usr/lib/python3.6/urllib/request.py", line 504, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.6/urllib/request.py", line 650, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
codalab.client.json_api_client.JsonApiException: Unable to update worksheet: Internal Server Error - Unexpected Internal Error ('latin-1' codec can't encode characters in position 7-19: ordinal not in range(256)). The administrators have been notified.
 (exit code 1, expected 0)
```
It appears that 908ce5c224ba75105aef1ceea0e73613bf812bd3 fixed it, although I'm not sure whether this is the best solution and whether this will work in production -- will need to test.

epicfaace committed:
fix: use integer division

epicfaace committed:
fix: don't encode strings when returning uid

epicfaace committed:
fix: fix invalid sqlite url

epicfaace committed:
fix: fix binary/unicode issues for netcat tests

epicfaace committed:
chore: black format

epicfaace committed:
fix: re-add truncating of console output

epicfaace committed:
chore: remove double parenthesis around print statements

epicfaace committed:
fix: properly truncate output

epicfaace committed:
chore: remove duplicate import

epicfaace committed:
chore: keep iterators through .items()

epicfaace committed:
chore: remove duplicate imports

epicfaace committed:
chore: change list(range to range

epicfaace committed:
chore: convert most list()'s to iterators

epicfaace committed:
fix: syntax error

epicfaace committed:
chore: remove more iter's

epicfaace committed:
chore: remove extra print statement

epicfaace committed:
chore: remove flush

epicfaace committed:
fix: replace eval(input()) with input()

epicfaace committed:
Use binary string

epicfaace committed:
Update tests/worker/file_util_test.py

epicfaace committed:
Update worker/codalabworker/worker.py

epicfaace committed:
Update worker/codalabworker/local_run/local_run_manager.py

epicfaace committed:
Update worker/codalabworker/local_run/local_dependency_manager.py

epicfaace committed:
Update test_cli.py

epicfaace committed:
Update worker/codalabworker/file_util.py

epicfaace committed:
Update setup.py

epicfaace committed:
Update worker/setup.py

epicfaace committed:
chore: replace list() with iterators

epicfaace committed:
fix: force python 3.6

epicfaace committed:
fix: store login cookie as json, not a pickled object

This is mainly due to problems with Python 3 reading existing Python
2-pickled cookies. Additionally, there are security issues with storing
pickles as cookies and bottle recommends us to store JSON objects anyway
(https://github.com/bottlepy/bottle/issues/900).

epicfaace commented:
@percyliang did you try this out yourself -- do you think it's ready to merge?

percyliang reviewed (commented):
Great to see this PR coming together - it will be so much nicer to be on Python 3!

I still need to try out the code and test drive it more thoroughly - since this is a big change, I'd want to make sure everything's okay.

Concretely, the main thing I'm concerned about is the handling of strings and unicode, which has been a huge issue with the existing code.  With Python 3 and everything being unicode, things should be much nicer, but there are still a bunch of byte-string conversions that exist at various points in the code.  Could you summarize briefly what the contract is for where things are bytes and where things are strings and where conversions need to happen?

percyliang commented:
Just tried testing the system with unicode and found the following problems:
- If I add '你好' into the worksheet markdown (either using `cl wedit` or in the web browser), it shows up as '??'.
- If I upload a file with unicode, `cl cat` works but `summarize_file` fails with "Unexpected Internal Error ('ascii' codec can't decode byte 0xe8 in position 0: ordinal not in range(128)). The administrators have been notified." 
- `cl wedit -T 你好` throws an exception
- `cl edit <bundle> -d 你好` also saves as '??'
Probably should have tests for these cases.

epicfaace commented:
@percyliang In each of the above cases, should codalab allow or not allow non-ascii characters like "你好"?

epicfaace commented:
For example, it seems like a lot of fields in the schema (such as worksheet title, tags) are supposed to be ascii -- why was this this case? Do we want to change it to allow unicode? https://github.com/codalab/codalab-worksheets/blob/54ab716491b774fd66a7172120c21bea543360b3/codalab/rest/schemas.py#L119

epicfaace commented:
> * If I upload a file with unicode, `cl cat` works but `summarize_file` fails with "Unexpected Internal Error ('ascii' codec can't decode byte 0xe8 in position 0: ordinal not in range(128)). The administrators have been notified."

How do you run `summarize_file`? Is there a way to run it from the command line?

percyliang commented:
Ideally we should allow unicode everywhere.  Not sure why the validate_ascii was there - probably because unicode wasn't working before.

epicfaace committed:
Merge

epicfaace committed:
fix: run tests with right version of python

epicfaace committed:
fix: print test output to stdout in real time

epicfaace committed:
chore: reformat black

epicfaace committed:
ci: set travis python encoding to utf8

epicfaace committed:
fix:  run competitiond with python3.6

epicfaace committed:
fix: print ascii for travis

epicfaace committed:
Merge remote-tracking branch 'origin/master' into python3

epicfaace committed:
fix (unicode): properly encode subprocess args before sending to command line

epicfaace committed:
fix: make sure fileobj passed to _upload_with_chunked_encoding is bytes

epicfaace committed:
fix: remove unicode restrictions

epicfaace committed:
fix: remove unicode_constructor

epicfaace committed:
Revert "fix: remove unicode_constructor"

This reverts commit ced5644aed5edc8c08e7bf6c5d1b7a86ab836921.

epicfaace committed:
update description for unicode_constructor

epicfaace committed:
fix: print only ascii in tests

epicfaace committed:
fix: properly accept unicode arguments on the command line

epicfaace committed:
chore: doc reformat

epicfaace committed:
fix: allow unicode in metadata

epicfaace committed:
Merge

epicfaace committed:
fix: add more unicode tests

epicfaace committed:
fix: don't require ascii on frontend

epicfaace committed:
fix: save test file in utf-8 encoding

epicfaace committed:
fix: make sure python encoding for server file is utf-8

epicfaace committed:
fix: open file correctly with with statement

epicfaace committed:
fix: allow unicode bodies

epicfaace committed:
test: add wedit tests for unicode

epicfaace committed:
remove non-ascii error handling

epicfaace committed:
reformat with black

epicfaace committed:
Merge remote-tracking branch 'origin/master' into python3

epicfaace committed:
fix: use python3 pip

epicfaace committed:
Revert "fix: use python3 pip"

This reverts commit 6ce7b316ce437a465e82dd7b4fcb85a41f3d3776.

epicfaace committed:
fix: change docs to be python 3 again

epicfaace committed:
use python3 pip

epicfaace committed:
try ENV LANG for python encoding

epicfaace committed:
remove PYTHONIOENCODING from travis env

epicfaace committed:
3.5 to 3.6

epicfaace committed:
codalabworker requires 3.6

epicfaace committed:
Merge

epicfaace committed:
fix: fix print

epicfaace committed:
black format

epicfaace committed:
fix typo

epicfaace committed:
fix: upgrade sqlalchemy

epicfaace committed:
fix: package_data values should be array

epicfaace committed:
downgrade sqlalchemy

epicfaace committed:
add test for unicode body

epicfaace committed:
Merge remote-tracking branch 'origin/master' into python3

epicfaace committed:
use 2to3 on gen-cli-docs

epicfaace committed:
fix gen-rest-docs and gen-cli-docs

epicfaace committed:
remove collation from mysql url

epicfaace committed:
add bundle model test for saving metadata in unicode

epicfaace committed:
black format

epicfaace committed:
test: test actual mysql and sqlite bundle models

epicfaace committed:
fix bundlemodel test

epicfaace committed:
fix: actually fix bundle_model_test

epicfaace committed:
Merge

epicfaace committed:
fix: use utf8mb4_unicode_ci collation

epicfaace committed:
specify collation properly

epicfaace committed:
fix: fix collation for worksheet item value and tag names

epicfaace committed:
black format

epicfaace committed:
Update worker/codalabworker/rest_client.py

epicfaace committed:
Merge

epicfaace committed:
use unicode and unicodetext

epicfaace committed:
black format

epicfaace committed:
fix: decode strings in hacky way

epicfaace committed:
fix: import array

epicfaace committed:
fix: fix worksheet_tags test

epicfaace committed:
Merge

epicfaace committed:
remove sqlite test for now

epicfaace committed:
remove encode and decode

epicfaace committed:
fix: go back to python-2 way

epicfaace committed:
fix: get a string branch

epicfaace committed:
replace python with python3.6

epicfaace committed:
remove comments

epicfaace committed:
fix: revert bundle_model tests, we  don't actually need them

epicfaace committed:
show help with "cl" argument, fixes #1332

epicfaace committed:
fix: allow unicode in worksheet tags

epicfaace committed:
test: add test for unicode in worksheet title

epicfaace committed:
remove bundle_model_test

epicfaace committed:
fix: unicode in metadata

epicfaace committed:
black format

epicfaace committed:
Revert tables.py

epicfaace committed:
attempt to add unicodecoerce custom types

epicfaace committed:
remove metadata_row check

epicfaace committed:
remove encode_str and decode_str

epicfaace committed:
re-add bundle_model_test

epicfaace committed:
add bundlemodeltest fix indent

epicfaace committed:
remove unicode support for title, description, tags

epicfaace committed:
fix --title wedit test

epicfaace committed:
add decode_str

epicfaace committed:
black format

epicfaace committed:
Remove if bytes statement

epicfaace committed:
Open json files using r mode instead of rb mode

epicfaace committed:
fix: remove unicode argument hack, because default encoding has been properly set in Dockerfile

epicfaace committed:
fix: remove unneeded try-except

epicfaace committed:
support binary args in tests

epicfaace committed:
remove unneeded try-catch

epicfaace committed:
black format

epicfaace committed:
Delete types.py

epicfaace committed:
Merge branch 'python3' of github.com:codalab/codalab-worksheets into python3

epicfaace committed:
Merge

epicfaace committed:
black format

epicfaace committed:
fix python -> python3.6

epicfaace committed:
fix gen-rest-docs

percyliang reviewed (approved):
This is great!  Remember to increment the version to 0.4.0.

epicfaace commented:
Why 0.4.0 (as opposed to, say, 1.0.0)?

epicfaace committed:
Merge

percyliang commented:
I was thinking `0.4.0` since it's the same functionality and is backwards compatible...

epicfaace commented:
I'll increment the version in another PR.
